sycl/include/sycl/detail/defines.hpp

```diff
@@ -27,7 +27,7 @@
 #define __SYCL_SPECIAL_CLASS
 #endif
 
-#if __has_cpp_attribute(__sycl_detail__::sycl_type)
+#if __has_cpp_attribute(__sycl_detail__::sycl_type) && (defined __SYCL_DEVICE_ONLY__)
```

Comment:
I wonder, can we update our attributes so `__has_cpp_attribute` returns `true` only for device compilation? Tagging @intel/dpcpp-cfe-reviewers

Reply:
I think you mean that *some* SYCL specific attributes can be enabled by the compiler only in SYCL device compilation mode and reported as unsupported in other modes (SYCL host or non-SYCL). Right?
I don't think we can change the semantic of `__has_cpp_attribute`.

Reply:
> I think you mean that some SYCL specific attributes can be enabled by the compiler only in SYCL device compilation mode and reported as unsupported in other modes (SYCL host or non-SYCL). Right?

Yes, thanks, that's what I meant. I would like us to simply rely on `__has_cpp_attibute` alone, without need to check for `__SYCL_DEVICE_ONLY__`.

Reply:
You can modify attribute definition to enable it only on device. You can silence the ignored attribute on host which is how we handle other device only attributes

Reply:
We will still have checks for attribute arg validity, see https://github.com/intel/llvm/pull/7881#issuecomment-1369699163

Reply:
Hmm...technically that means attribute semantics is incorrect i.e. it is invalid code. I cannot recall off the top of my head how we handle this usually. @smanna12 @AaronBallman should we diagnose invalid attribute usage when it is ignored on device? I vaguely remember a discussion about this once but I cannot recall the resolution

Reply:
Typically, when an attribute is ignored, we warn about the "unknown" attribute and that is it. e.g., [[no_unique_address]] is a standard attribute that accepts no arguments, but we don't support it on Windows because of <reasons>, and so: https://godbolt.org/z/6Yh1jWx89

Reply:
Ok Thank you for confirming.

Reply:
@romanovvlad if this attribute is not required on host, you can modify the attribute definition as well as silence other diagnostics by not handling the attribute on host as you have already done. I think we can then remove __SYCL_DEVICE_ONLY__ check as @AlexeySachkov suggested.

Reply:
I spend a bit of time looking into this. __SYCL_DEVICE_ONLY__ check cannot be removed at the moment because `__has_attribute` does not consider LangOpts when generating attributes in TableGen. This should be fixed in upstream clang if possible. We can do this as a separate task IMO.